packages/gatsby-transformer-remark/src/extend-node-type.js

```diff
@@ -144,30 +144,30 @@ module.exports = (
 
     async function getAST(markdownNode) {
       const cacheKey = astCacheKey(markdownNode)
+
+      const promise = ASTPromiseMap.get(cacheKey)
+      if (promise) {
+        // We are already generating AST, so let's wait for it
+        return promise
+      }
+
       const cachedAST = await cache.get(cacheKey)
       if (cachedAST) {
         return cachedAST
-      } else if (ASTPromiseMap.has(cacheKey)) {
-        // We are already generating AST, so let's wait for it
-        return await ASTPromiseMap.get(cacheKey)
-      } else {
-        const ASTGenerationPromise = getMarkdownAST(markdownNode)
-        ASTGenerationPromise.then(markdownAST => {
-          ASTPromiseMap.delete(cacheKey)
-          return cache.set(cacheKey, markdownAST)
-        }).catch(err => {
-          ASTPromiseMap.delete(cacheKey)
-          return err
-        })
-        // Save new AST to cache and return
-        // We can now release promise, as we cached result
-        ASTPromiseMap.set(cacheKey, ASTGenerationPromise)
-        return ASTGenerationPromise
       }
+
+      const ASTGenerationPromise = getMarkdownAST(markdownNode)
+      ASTPromiseMap.set(cacheKey, ASTGenerationPromise)
+
+      // Return the promise that will cache the result if good, and deletes the promise from local cache either way
+      // Note that if this cache hits for another parse, it won't have to wait for the cache
+      return ASTGenerationPromise.then(markdownAST =>
+        // Return a promise that waits for the cache to be set, but that does return the generated AST
+        cache.set(cacheKey, markdownAST).then(() => markdownAST)
+      ).finally(() => ASTPromiseMap.delete(cacheKey))
```

Comment:
Can we use `await` and `try fnally`? It is a bit inconsistent this way.

Reply:
yeah good catch